source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -138,10 +146,39 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void ProxyFilter::onLoadDnsCacheComplete(const Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
-  ENVOY_STREAM_LOG(debug, "load DNS cache complete, continuing", *decoder_callbacks_);
+void ProxyFilter::addHostAddressToFilterState(
+    const Network::Address::InstanceConstSharedPtr& address) {
+  ENVOY_STREAM_LOG(trace, "Adding resolved host {} to filter state", *decoder_callbacks_,
+                   address->asString());
+  const Envoy::StreamInfo::FilterStateSharedPtr& filter_state =
+      decoder_callbacks_->streamInfo().filterState();
+
+  if (!filter_state->hasData<StreamInfo::AddressSetAccessor>(
+          StreamInfo::KEY_DYNAMIC_PROXY_UPSTREAM_ADDR)) {
+    auto address_set = std::make_unique<StreamInfo::AddressSetAccessorImpl>();
+    address_set->add(address);
+
+    filter_state->setData(StreamInfo::KEY_DYNAMIC_PROXY_UPSTREAM_ADDR, std::move(address_set),
```

Comment:
I'd actually like @mattklein123 's thoughts on this.
I'd mildly prefer to have rbac just look up the ip:host mapping to copying it all into stream info for every request, but that'd be a hard dep on the dynamic proxy code which may be worse in the long run.

Reply:
If it were up to me I would probably leave this as is, but config guard tracking this in filter state to avoid the perf hit for people that don't care.